File: electron/agent/tools/memory.ts
/** memory tool — semantic + full-text search over the workspace RAG index,
 *  fused with the global knowledge-sources index (filtered to sources
 *  enabled for this workspace). Verifies RAG, opens both stores, resolves
 *  the embedder that built each index, embeds the query, and merges
 *  vector + FTS top-K via reciprocal rank fusion (RRF, k=60). Knowledge
 *  hits are labeled with their doc origin so the model can cite
 *  "docs.react.dev" vs repo files. Read-only and auto-approved. */

import { tool } from 'ai';
import { z } from 'zod';
import * as path from 'node:path';
import * as fs from 'node:fs';
import type { ToolContext } from './tool-context.js';
import { openRagStore, type VectorHit, type FtsHit } from '../../rag/store.js';
import { resolveForQuery } from '../../rag/resolve.js';
import { localModelExists } from '../../rag/local-onnx-embedder.js';
import { isRagCloudConfigured } from '../system-model.js';
import { hydrateRagConfig } from '../../configStore.js';
import * as workspaceStore from '../../store.js';
import { knowledgeDbPath, openKnowledgeStore } from '../../knowledge/store.js';
import type { Embedder } from '../../rag/embedder.js';
import type { ToolResult, ToolRegistration } from './types';

const DEFAULT_K = 5;
const MAX_K = 20;
/** RRF constant. Standard value from the original TREC paper; balances
 *  head vs tail of the rankings without tuning. */
const RRF_K = 60;

/** Knowledge hit decorated with its source's display name so results can cite the origin ("React Docs · react.dev/guide") distinctly from repo files. */
type KnowledgeHit = (VectorHit | FtsHit) & { sourceName: string };

const NO_KNOWLEDGE: { hits: KnowledgeHit[]; total: number } = { hits: [], total: 0 };

/** Tagged preparation failures: the workspace half surfaces these fatally;
 *  the knowledge half treats any of them as "no knowledge results". */
class RagUnusableError extends Error {}
class QueryEmbedError extends Error {}

interface PreparedQuery {
  embedderId: string;
  queryVec: number[];
}

/** Shared body — testable without the SDK wrapper. ctx-free; the
 *  workspaceId comes from the caller (the SDK factory pulls it from
 *  ToolContext, tests pass it directly). */
export async function runMemory(
  query: string,
  k: number,
  workspaceId: string,
): Promise<ToolResult> {
  if (!query.trim()) {
    return { status: 'failed', output: 'Missing required arg: query' };
  }
  if (!workspaceId) {
    return {
      status: 'failed',
      output: 'No active workspace bound to this session.',
    };
  }

  // 1. Workspace gate governs only the workspace half — a workspace with
  //    RAG disabled still reaches registered knowledge sources.
  const wsEnabled =
    workspaceStore.listRagEnabledWorkspaces().includes(workspaceId);

  // 2. Open the workspace store. If this throws (sqlite-vec missing, db
  //    corrupt), surface the actual error rather than a generic "failed".
  let ragStore: ReturnType<typeof openRagStore> | null = null;
  if (wsEnabled) {
    try {
      ragStore = openRagStore(workspaceId);
    } catch (e: unknown) {
      const msg = e instanceof Error ? e.message : String(e);
      return { status: 'failed', output: `Failed to open RAG index: ${msg}` };
    }
  }

  try {
    const wsTotal = ragStore?.chunkCount() ?? 0;

    // 3. Query preparation (embedder resolution + query embedding) is lazy
    //    and memoized: paid at most once, and never started when no half
    //    ends up needing it. Failures are tagged so the workspace half can
    //    surface them fatally while the knowledge half swallows them.
    const kClamped = Math.min(Math.max(k, 1), MAX_K);
    let prepared: PreparedQuery | null = null;
    const prepareQuery = async (): Promise<PreparedQuery> => {
      if (!prepared) {
        let embedder: Embedder;
        let embedderId: string;
        try {
          const ws = workspaceStore.listWorkspaces().find((w) => w.id === workspaceId);
          const ragConfig = hydrateRagConfig(ws?.ragConfig);
          ({ embedder, embedderId } = resolveForQuery({
            config: ragConfig,
            localAvailable: localModelExists(),
            cloudConfigured: isRagCloudConfigured(),
          }));
        } catch (e: unknown) {
          throw new RagUnusableError(
            e instanceof Error ? e.message : String(e),
          );
        }
        try {
          const vecs = await embedder.embed([query]);
          prepared = { embedderId, queryVec: vecs[0] };
        } catch (e: unknown) {
          throw new QueryEmbedError(
            e instanceof Error ? e.message : String(e),
          );
        }
      }
      return prepared;
    };

    if (wsEnabled) {
      try {
        await prepareQuery();
      } catch (e: unknown) {
        if (e instanceof RagUnusableError) {
          return { status: 'failed', output: `RAG index unusable: ${e.message}` };
        }
        return { status: 'failed', output: `Embedding query failed: ${(e as Error).message}` };
      }
    }

    let wsFused: Array<VectorHit | FtsHit> = [];
    if (ragStore && wsTotal > 0) {
      const { queryVec } = await prepareQuery();
      wsFused = fuse(
        ragStore.queryByVector(queryVec, kClamped),
        ragStore.queryByFts(query, kClamped),
        kClamped,
      );
    }

    const knowledge = await searchKnowledgeSources({
      workspaceId,
      query,
      k: kClamped,
      prepareQuery,
    });

    // 5. Reciprocal rank fusion: within each store first, then across stores.
    const fused = fuse(wsFused, knowledge.hits, kClamped);
    const grandTotal = wsTotal + knowledge.total;

    if (fused.length === 0) {
      if (!wsEnabled) {
        return {
          status: 'executed',
          output:
            `RAG is not enabled for this workspace. ` +
            `Enable it in Settings → Memory & RAG (toggles the Switch on for this workspace; ` +
            `ingestion will run automatically on first enable).`,
        };
      }
      if (wsTotal === 0) {
        return {
          status: 'executed',
          output:
            `RAG index for this workspace is empty. ` +
            `Re-trigger ingestion from Settings → Memory & RAG → Re-index.`,
        };
      }
      return {
        status: 'executed',
        output: `No matches for "${query}" across ${grandTotal} indexed chunks.`,
      };
    }

    // 6. Format. Truncate body to BODY_CAP chars to keep the tool result
    //    readable and avoid bloating the model's context.
    const BODY_CAP = 1500;
    const lines = fused.map((hit, i) => {
      const loc = 'sourceName' in hit
        ? `[${hit.sourceName}] ${hit.path}`
        : `${shortPath(hit.path)}:${hit.startLine}` +
          (hit.symbol ? ` (${hit.symbol})` : '');
      const sim = 'similarity' in hit
        ? ` · ${Math.round(hit.similarity * 100)}%`
        : '';
      const body = hit.content.length > BODY_CAP
        ? hit.content.slice(0, BODY_CAP) + '\n…[truncated]'
        : hit.content;
      return `[${i + 1}] ${loc}${sim}\n${body}`;
    });

    const text =
      `Found ${fused.length} relevant chunk${fused.length === 1 ? '' : 's'} ` +
      `for "${query}" (out of ${grandTotal}):\n\n${lines.join('\n\n')}`;

    return {
      status: 'executed',
      output: text,
      // `display.kind:'text'` makes ToolCallCard render the body in a collapsible section (like web_search/web_fetch). Without it the card shows only the header — the user sees the tool was called but can't read what it found.
      display: { kind: 'text', text },
    };
  } finally {
    ragStore?.close();
  }
}

/** Search the global knowledge-sources index. Best-effort by design: any
 *  failure here (db missing/corrupt, registry error) degrades to "no
 *  knowledge results" so a broken global DB never fails the tool. */
async function searchKnowledgeSources(opts: {
  workspaceId: string;
  query: string;
  k: number;
  prepareQuery: () => Promise<PreparedQuery>;
}): Promise<{ hits: KnowledgeHit[]; total: number }> {
  try {
    const dbPath = knowledgeDbPath();
    // existsSync guard: openRagStoreAt would otherwise create an empty db as a side effect of every query.
    if (!fs.existsSync(dbPath)) return NO_KNOWLEDGE;
    const ks = openKnowledgeStore(dbPath);
    try {
      const { embedderId, queryVec } = await opts.prepareQuery();
      // First-embedder-wins pinning (meta.embedderId): silently skip on
      // mismatch — unlike the workspace path, which surfaces an error.
      const pinned = ks.rag.getMeta('embedderId');
      if (pinned && pinned !== embedderId) return NO_KNOWLEDGE;

      const enabledIds = new Set(ks.enabledSourceIdsFor(opts.workspaceId));
      const sources = ks.listSources();
      // Total counts only what this workspace can see — chunks behind
      // disabled sources must not leak into the reported coverage.
      const visibleChunks = sources
        .filter((s) => enabledIds.has(s.id))
        .reduce((n, s) => n + s.chunkCount, 0);
      if (enabledIds.size === 0 || visibleChunks === 0) return NO_KNOWLEDGE;

      const names = new Map(ks.listSources().map((s) => [s.id, s.name] as const));
      // Over-fetch then filter by sourceId: cheap post-filtering beats
      // sqlite-vec metadata-filter complexity (plan decision 4).
      const overFetch = opts.k * 3;
      const isEnabled = (
        h: VectorHit | FtsHit,
      ): h is (VectorHit | FtsHit) & { sourceId: string } =>
        h.sourceId != null && enabledIds.has(h.sourceId);
      const kVec = ks.rag.queryByVector(queryVec, overFetch).filter(isEnabled);
      const kFts = ks.rag.queryByFts(opts.query, overFetch).filter(isEnabled);
      const hits: KnowledgeHit[] = fuse(kVec, kFts, opts.k).map((h) => ({
        ...h,
        sourceName: names.get(h.sourceId) ?? h.sourceId,
      }));
      return { hits, total: visibleChunks };
    } finally {
      // A close() failure here must not discard already-computed hits.
      try {
        ks.close();
      } catch {
        /* nothing actionable */
      }
    }
  } catch (e) {
    console.warn(
      '[memory] knowledge search skipped:',
      e instanceof Error ? e.message : String(e),
    );
    return NO_KNOWLEDGE;
  }
}

/** Reciprocal Rank Fusion — zero-parameter merge of two rankings using rank-only signals; generic over distinct hit types so VectorHit + FtsHit fuse without forcing one score shape. */
function fuse<T1 extends { id: string }, T2 extends { id: string }>(
  vec: T1[],
  fts: T2[],
  k: number,
): Array<T1 | T2> {
  type Item = T1 | T2;
  const scores = new Map<string, { item: Item; score: number }>();
  for (let i = 0; i < vec.length; i++) {
    scores.set(vec[i].id, { item: vec[i], score: 1 / (RRF_K + i + 1) });
  }
  for (let i = 0; i < fts.length; i++) {
    const s = 1 / (RRF_K + i + 1);
    const existing = scores.get(fts[i].id);
    if (existing) existing.score += s;
    else scores.set(fts[i].id, { item: fts[i], score: s });
  }
  return [...scores.values()]
    .sort((a, b) => b.score - a.score)
    .slice(0, k)
    .map((s) => s.item);
}

/** Workspace-relative path for compact display. Falls back to basename
 *  if the path isn't under the workspace (e.g. temp fixture in tests). */
function shortPath(absPath: string): string {
  const parts = absPath.split(/[/\\]/).filter(Boolean);
  if (parts.length <= 3) return absPath;
  return '…/' + parts.slice(-2).join('/');
}

/** Re-export the hit types so consumers can import everything from this
 *  module without reaching into store.js. */
export type { VectorHit, FtsHit };

// ─── SDK factory (the registration path used by orchestrator-sdk) ──────

export function createMemoryTool(ctx: ToolContext) {
  return tool({
    description:
      'FIRST tool to call for ANY codebase question. Searches the workspace RAG index ' +
      'and registered knowledge sources by meaning and returns ranked chunks in ~0.5s. ' +
      'Call this BEFORE directory_tree, list_dir, read_file, or grep. Returns file path + ' +
      'line range + source body for each match; knowledge-source hits are labeled [source] origin ' +
      '(e.g. [React Docs] react.dev/learn) — cite that origin when using them. ' +
      'Example: memory({ query: "how is authentication handled" }) → returns the auth files + code.',
    inputSchema: z.object({
      query: z
        .string()
        .describe(
          'Natural language query describing what you are looking for. Examples: ' +
          '"user authentication flow", "database connection setup", "API route definitions".',
        ),
      k: z
        .number()
        .int()
        .min(1)
        .max(MAX_K)
        .optional()
        .describe(`Top-K chunks to return. Default ${DEFAULT_K}, max ${MAX_K}.`),
    }),
    execute: async ({ query, k }) =>
      runMemory(query, k ?? DEFAULT_K, ctx.workspaceId),
  });
}

// ─── Legacy ToolRegistration shape (kept for the registry's non-SDK map) ──
// Consumed by the legacy orchestrator (USE_SDK_ORCHESTRATOR=false). Today that path is dormant, but the registry imports it for shape parity + so a flip back doesn't break; the execute signature mirrors the SDK shape.

export const memoryTool: ToolRegistration = {
  name: 'memory' as const,
  definition: {
    name: 'memory' as const,
    description:
      'FIRST tool to call for ANY codebase question. Searches the workspace RAG index ' +
      'and registered knowledge sources by meaning and returns ranked chunks in ~0.5s. ' +
      'Call BEFORE directory_tree, list_dir, read_file, or grep. Returns file path + line ' +
      'range + source body; knowledge-source hits are labeled [source] origin.',
    input_schema: {
      type: 'object' as const,
      properties: {
        query: { type: 'string', description: 'Natural language: "how is authentication handled", "database setup", "API routes".' },
        k: { type: 'number', description: `Top-K results. Default ${DEFAULT_K}, max ${MAX_K}.` },
      },
      required: ['query'],
    },
  },
  riskTier: 'read_only',
  requiresWorktree: false,
  timeoutMs: 5_000,
  autoApproveIn: ['plan', 'ask', 'edit', 'full'],
  execute: async (args, ctx) => {
    const query = typeof args.query === 'string' ? args.query : '';
    const k = typeof args.k === 'number' ? args.k : DEFAULT_K;
    // The legacy ToolContext in ./types doesn't carry workspaceId (only
    // the SDK one does). The legacy path is dormant today
    // (USE_SDK_ORCHESTRATOR=true), so a missing workspaceId here just
    // surfaces an actionable hint via runMemory.
    return runMemory(query, k, (ctx as { workspaceId?: string }).workspaceId ?? '');
  },
};

// path import is only used by shortPath above; keep it here so the
// import isn't shaken out by tree-shaking in some bundler configs.
void path;

File: src-tauri/crates/tide-tools/src/tools/memory.rs
//! memory — port of `app/core/agent/tools/memory.ts` (): semantic +
//! full-text search over the workspace RAG index, fused with the global
//! knowledge-sources index, merged via reciprocal rank fusion (RRF, k=60).
//!
//! The TS tool opened SQLite + sqlite-vec stores directly. Rust keeps the
//! store behind the [`MemoryIndex`] seam: tide-rag implements it once the
//! embedding/ingestion pipeline lands; until then the orchestrator registers
//! the tool with no index and queries return the TS-faithful "RAG is not
//! enabled" hint. The knowledge-source half of the TS fusion (a global
//! index filtered per workspace) is the implementor's concern — hits carry
//! [`MemoryHit::source_name`] so origins stay citable either way.

use std::sync::Arc;

use serde::{Deserialize, Serialize};
use serde_json::json;

use crate::permission::RiskTier;
use crate::{Tool, ToolContext, ToolDisplay, ToolError, ToolOutcome, ToolSpec};

use super::arg_str;

const DESCRIPTION: &str = "FIRST tool to call for ANY codebase question. Searches the workspace RAG index and registered knowledge sources by meaning and returns ranked chunks in ~0.5s. Call BEFORE directory_tree, list_dir, read_file, or grep. Returns file path + line range + source body; knowledge-source hits are labeled [source] origin.";

const DEFAULT_K: u64 = 5;
const MAX_K: u64 = 20;
/// RRF constant. Standard value from the original TREC paper; balances
/// head vs tail of the rankings without tuning.
const RRF_K: f64 = 60.0;
/// Body truncation cap — keeps the tool result readable and the model's
/// context unbloated.
const BODY_CAP: usize = 1500;

/// One ranked chunk. Workspace hits carry path/line/symbol; knowledge hits
/// additionally carry `source_name` so results can cite the origin
/// ("React Docs · react.dev/guide") distinctly from repo files.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MemoryHit {
    pub id: String,
    pub path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub symbol: Option<String>,
    pub start_line: u64,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub similarity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_name: Option<String>,
}

/// The search backend the memory tool consults — the seam tide-rag fills
/// in (vector + FTS rankings over the workspace index and any registered
/// knowledge sources visible to the workspace). Methods take the
/// workspace id because the TS tool routed every store operation through
/// it — one process-wide backend resolves both halves per workspace.
pub trait MemoryIndex: std::fmt::Debug + Send + Sync {
    /// Total indexed chunks visible to this workspace.
    fn total_chunks(&self, workspace_id: &str) -> u64;
    /// Top-k vector (semantic) ranking for the query.
    fn vector_hits(&self, workspace_id: &str, query: &str, k: usize) -> Vec<MemoryHit>;
    /// Top-k full-text ranking for the query.
    fn fts_hits(&self, workspace_id: &str, query: &str, k: usize) -> Vec<MemoryHit>;
}

/// Process-wide backend slot (the [`TodoState::shared`] pattern): the
/// command layer installs the tide-rag-backed index once the RAG domain
/// comes online; until then queries take the "RAG is not enabled" hint.
static SHARED_INDEX: std::sync::RwLock<Option<std::sync::Arc<dyn MemoryIndex>>> =
    std::sync::RwLock::new(None);

/// Install (or clear) the process-wide memory index backend.
pub fn set_shared_memory_index(index: Option<std::sync::Arc<dyn MemoryIndex>>) {
    let mut guard = SHARED_INDEX.write().expect("memory index slot poisoned");
    *guard = index;
}

/// The installed backend, if any.
pub fn shared_memory_index() -> Option<std::sync::Arc<dyn MemoryIndex>> {
    SHARED_INDEX
        .read()
        .expect("memory index slot poisoned")
        .clone()
}

/// Reciprocal Rank Fusion — zero-parameter merge of two rankings using
/// rank-only signals; generic id-keyed so vector + FTS hits fuse without
/// forcing one score shape. Port of the TS `fuse`.
pub fn rrf_fuse(vec: Vec<MemoryHit>, fts: Vec<MemoryHit>, k: usize) -> Vec<MemoryHit> {
    let mut scores: Vec<(f64, MemoryHit)> = Vec::with_capacity(vec.len() + fts.len());
    let mut index_of: std::collections::HashMap<String, usize> =
        std::collections::HashMap::with_capacity(vec.len() + fts.len());
    for (rank, hit) in vec.into_iter().enumerate() {
        index_of.insert(hit.id.clone(), scores.len());
        scores.push((1.0 / (RRF_K + rank as f64 + 1.0), hit));
    }
    for (rank, hit) in fts.into_iter().enumerate() {
        let s = 1.0 / (RRF_K + rank as f64 + 1.0);
        match index_of.get(&hit.id) {
            Some(&i) => scores[i].0 += s,
            None => {
                index_of.insert(hit.id.clone(), scores.len());
                scores.push((s, hit));
            }
        }
    }
    scores.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));
    scores.truncate(k);
    scores.into_iter().map(|(_, hit)| hit).collect()
}

/// Workspace-relative path for compact display. Falls back to the full
/// path when it isn't deep enough to shorten (e.g. temp fixture paths).
pub(crate) fn short_path(abs_path: &str) -> String {
    let parts: Vec<&str> = abs_path
        .split(['/', '\\'])
        .filter(|p| !p.is_empty())
        .collect();
    if parts.len() <= 3 {
        return abs_path.to_string();
    }
    format!("…/{}", parts[..].split_at(parts.len() - 2).1.join("/"))
}

/// Shared body — testable without the trait object wrapper; the
/// workspace_id comes from the caller (the tool pulls it from ToolContext).
pub(crate) fn run_memory(
    query: &str,
    k: u64,
    workspace_id: &str,
    index: Option<&dyn MemoryIndex>,
) -> ToolOutcome {
    if query.trim().is_empty() {
        return ToolOutcome::failed("Missing required arg: query");
    }
    if workspace_id.is_empty() {
        return ToolOutcome::failed("No active workspace bound to this session.");
    }

    // No index wired ≈ the TS "RAG not enabled for this workspace" gate —
    // the tool stays useful (actionable hint) instead of failing.
    let Some(index) = index else {
        return ToolOutcome::executed(
            "RAG is not enabled for this workspace. Enable it in Settings → Memory & RAG (toggles the Switch on for this workspace; ingestion will run automatically on first enable).",
        );
    };

    let total = index.total_chunks(workspace_id);
    if total == 0 {
        return ToolOutcome::executed(
            "RAG index for this workspace is empty. Re-trigger ingestion from Settings → Memory & RAG → Re-index.",
        );
    }

    let k_clamped = k.clamp(1, MAX_K) as usize;
    let fused = rrf_fuse(
        index.vector_hits(workspace_id, query, k_clamped),
        index.fts_hits(workspace_id, query, k_clamped),
        k_clamped,
    );

    if fused.is_empty() {
        return ToolOutcome::executed(format!(
            "No matches for \"{query}\" across {total} indexed chunks."
        ));
    }

    let lines = fused
        .iter()
        .enumerate()
        .map(|(i, hit)| {
            let loc = match &hit.source_name {
                Some(source) => format!("[{source}] {}", hit.path),
                None => format!(
                    "{}:{}{}",
                    short_path(&hit.path),
                    hit.start_line,
                    hit.symbol
                        .as_deref()
                        .map(|s| format!(" ({s})"))
                        .unwrap_or_default()
                ),
            };
            let sim = hit
                .similarity
                .map(|s| format!(" · {}%", (s * 100.0).round() as u64))
                .unwrap_or_default();
            let body = if hit.content.chars().count() > BODY_CAP {
                let mut cut: String = hit.content.chars().take(BODY_CAP).collect();
                cut.push_str("\n…[truncated]");
                cut
            } else {
                hit.content.clone()
            };
            format!("[{}] {loc}{sim}\n{body}", i + 1)
        })
        .collect::<Vec<_>>();

    let text = format!(
        "Found {} relevant chunk{} for \"{query}\" (out of {total}):\n\n{}",
        fused.len(),
        if fused.len() == 1 { "" } else { "s" },
        lines.join("\n\n")
    );

    ToolOutcome::executed(text.clone()).with_display(ToolDisplay::Text { text })
}

pub struct MemoryTool {
    index: Option<Arc<dyn MemoryIndex>>,
}

impl MemoryTool {
    pub fn new(index: Option<Arc<dyn MemoryIndex>>) -> Self {
        Self { index }
    }
}

impl Tool for MemoryTool {
    fn spec(&self) -> ToolSpec {
        ToolSpec {
            name: "memory".into(),
            description: DESCRIPTION.into(),
            parameters: json!({
                "type": "object",
                "properties": {
                    "query": { "type": "string", "description": "Natural language: \"how is authentication handled\", \"database setup\", \"API routes\"." },
                    "k": { "type": "number", "description": "Top-K results. Default 5, max 20." }
                },
                "required": ["query"]
            }),
        }
    }

    fn risk_tier(&self) -> RiskTier {
        RiskTier::ReadOnly
    }

    fn execute(
        &self,
        ctx: &ToolContext,
        args: serde_json::Value,
    ) -> Result<ToolOutcome, ToolError> {
        let query = arg_str(&args, "query");
        let k = args.get("k").and_then(|v| v.as_u64()).unwrap_or(DEFAULT_K);
        // The constructor-bound index wins (tests); production rides the
        // process-wide slot installed by the RAG command layer.
        let shared = shared_memory_index();
        let index = self.index.as_deref().or(shared.as_deref());
        Ok(run_memory(&query, k, &ctx.workspace_id, index))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::OutcomeStatus;

    #[derive(Debug)]
    struct FakeIndex {
        total: u64,
        vector: Vec<MemoryHit>,
        fts: Vec<MemoryHit>,
    }

    impl MemoryIndex for FakeIndex {
        fn total_chunks(&self, _workspace_id: &str) -> u64 {
            self.total
        }
        fn vector_hits(&self, _workspace_id: &str, _query: &str, k: usize) -> Vec<MemoryHit> {
            self.vector.iter().take(k).cloned().collect()
        }
        fn fts_hits(&self, _workspace_id: &str, _query: &str, k: usize) -> Vec<MemoryHit> {
            self.fts.iter().take(k).cloned().collect()
        }
    }

    fn hit(id: &str, path: &str, symbol: Option<&str>, similarity: Option<f64>) -> MemoryHit {
        MemoryHit {
            id: id.into(),
            path: path.into(),
            symbol: symbol.map(str::to_owned),
            start_line: 10,
            content: format!("content of {id}"),
            similarity,
            source_name: None,
        }
    }

    #[test]
    fn missing_query_fails() {
        let out = run_memory("", 5, "ws1", None);
        assert_eq!(out.status, OutcomeStatus::Failed);
        assert_eq!(out.output, "Missing required arg: query");
    }

    #[test]
    fn missing_workspace_fails() {
        let out = run_memory("auth flow", 5, "", None);
        assert_eq!(out.status, OutcomeStatus::Failed);
        assert_eq!(out.output, "No active workspace bound to this session.");
    }

    #[test]
    fn no_index_reports_not_enabled_hint() {
        let out = run_memory("anything", 5, "ws1", None);
        assert_eq!(out.status, OutcomeStatus::Executed);
        assert!(out
            .output
            .starts_with("RAG is not enabled for this workspace."));
        assert!(out.output.contains("Settings → Memory & RAG"));
    }

    #[test]
    fn empty_index_reports_reindex_hint() {
        let index = FakeIndex {
            total: 0,
            vector: vec![],
            fts: vec![],
        };
        let out = run_memory("anything", 5, "ws1", Some(&index));
        assert_eq!(out.status, OutcomeStatus::Executed);
        assert!(out
            .output
            .starts_with("RAG index for this workspace is empty."));
    }

    #[test]
    fn no_matches_reports_total() {
        let index = FakeIndex {
            total: 42,
            vector: vec![],
            fts: vec![],
        };
        let out = run_memory("zzz", 5, "ws1", Some(&index));
        assert_eq!(out.status, OutcomeStatus::Executed);
        assert_eq!(
            out.output,
            "No matches for \"zzz\" across 42 indexed chunks."
        );
    }

    #[test]
    fn formats_workspace_and_knowledge_hits() {
        let index = FakeIndex {
            total: 7,
            vector: vec![hit("c1", "/repo/src/auth.ts", Some("login"), Some(0.87))],
            fts: vec![],
        };
        let out = run_memory("how does login work", 5, "ws1", Some(&index));
        assert_eq!(out.status, OutcomeStatus::Executed);
        assert!(out
            .output
            .starts_with("Found 1 relevant chunk for \"how does login work\" (out of 7):"));
        assert!(out
            .output
            .contains("[1] /repo/src/auth.ts:10 (login) · 87%\ncontent of c1"));
        assert!(matches!(out.display, Some(ToolDisplay::Text { .. })));
    }

    #[test]
    fn knowledge_hits_carry_source_label() {
        let knowledge = MemoryHit {
            source_name: Some("React Docs".into()),
            path: "react.dev/learn".into(),
            ..hit("k1", "react.dev/learn", None, None)
        };
        let index = FakeIndex {
            total: 3,
            vector: vec![knowledge],
            fts: vec![],
        };
        let out = run_memory("hooks", 5, "ws1", Some(&index));
        assert!(out.output.contains("[1] [React Docs] react.dev/learn"));
    }

    #[test]
    fn long_bodies_truncate_at_cap() {
        let long = MemoryHit {
            content: "y".repeat(BODY_CAP + 500),
            ..hit("c1", "/repo/a/b/c/d/long.ts", None, None)
        };
        let index = FakeIndex {
            total: 1,
            vector: vec![long],
            fts: vec![],
        };
        let out = run_memory("long", 5, "ws1", Some(&index));
        assert!(out.output.contains("…[truncated]"));
        assert!(!out.output.contains(&"y".repeat(BODY_CAP + 100)));
    }

    #[test]
    fn rrf_prefers_hits_present_in_both_rankings() {
        let both = hit("both", "/repo/both.ts", None, None);
        let vec = vec![hit("v-only", "/repo/v.ts", None, None), both.clone()];
        let fts = vec![both, hit("f-only", "/repo/f.ts", None, None)];
        let fused = rrf_fuse(vec, fts, 3);
        assert_eq!(fused[0].id, "both");
        assert_eq!(fused.len(), 3);
    }

    #[test]
    fn rrf_respects_k_cap() {
        let vec: Vec<_> = (0..10)
            .map(|i| hit(&format!("v{i}"), "/repo/x.ts", None, None))
            .collect();
        let fused = rrf_fuse(vec, vec![], 3);
        assert_eq!(fused.len(), 3);
    }

    #[test]
    fn short_path_keeps_shallow_paths_whole() {
        assert_eq!(short_path("/repo/a.ts"), "/repo/a.ts");
        assert_eq!(short_path("/very/deep/path/file.ts"), "…/path/file.ts");
        // Both separators split; the join is always forward-slash (TS join('/')).
        assert_eq!(short_path("C:\\a\\b\\c\\d.ts"), "…/c/d.ts");
    }

    #[test]
    fn execute_routes_through_trait_and_clamps_k() {
        let tmp = tempfile::tempdir().unwrap();
        let mut ctx = ToolContext::new(tmp.path());
        ctx.workspace_id = "ws1".into();
        let tool = MemoryTool::new(Some(Arc::new(FakeIndex {
            total: 9,
            vector: vec![hit("c1", "/repo/src/x.ts", Some("f"), Some(0.5))],
            fts: vec![],
        })));
        assert_eq!(tool.spec().name, "memory");
        assert_eq!(tool.risk_tier(), RiskTier::ReadOnly);

        // k=99 clamps to MAX_K (20) — the fake index only returns 1 hit.
        let out = tool
            .execute(&ctx, serde_json::json!({ "query": "x", "k": 99 }))
            .unwrap();
        assert_eq!(out.status, OutcomeStatus::Executed);
        assert!(out.output.contains("(out of 9)"));
    }
}
